[All hosts](security docs refresh) (#3010)
OfficeDev/office-js-docs-pr: docs/develop/sso-in-office-add-ins.md

---
title: Enable single sign-on (SSO) in an Office Add-in
description: 'Learn the key steps to enable single sign-on (SSO) for your Office Add-in using common Microsoft personal, work, or education accounts.'
ms.date: 01/07/2022
ms.localizationpriority: high
---

# Enable single sign-on (SSO) in an Office Add-in

Users sign in to Office (online, mobile, and desktop platforms) using either their personal Microsoft account or their Microsoft 365 Education or work account. You can take advantage of this and use single sign-on (SSO) to authenticate and authorize the user to your add-in without requiring the user to sign in a second time.

![An image showing the sign-in process for an add-in.](../images/sso-for-office-addins.png)

## How SSO works at runtime

The following diagram shows how the SSO process works. The blue elements represent Office or the Microsoft identity platform. The gray elements represent the code you write and include the client-side code (task pane) and the server-side code for your add-in.

![A diagram that shows the SSO process.](../images/sso-overview-diagram.svg)

> 1. In the add-in, your JavaScript code calls the Office.js API [getAccessToken](/javascript/api/office-runtime/officeruntime.auth#getAccessToken_options_). If the user is already signed in to Office, the Office host will return the access token with the claims of the signed in user.

Reviewer: ```suggestion
1. In the add-in, your JavaScript code calls the Office.js API [getAccessToken](/javascript/api/office-runtime/officeruntime.auth#getAccessToken_options_). If the user is already signed in to Office, the Office host returns the access token with the claims of the signed in user.
```

2. If the user is not signed in, the Office host application opens a dialog box for the user to sign in. Office redirects to the Microsoft identity platform to complete the sign-in process.
3. If this is the first time the current user has used your add-in, they are prompted to consent.
4. The Office host application requests the **access token** from the Microsoft identity platform for the current user.

> 5. The Microsoft identity platform returns the access token to Office. Office will cache the token on your behalf so that future calls to **getAccessToken** simply return the cached token.

Reviewer: ```suggestion
5. The Microsoft identity platform returns the access token to Office. Office caches the token on your behalf so that future calls to **getAccessToken** simply return the cached token.
```

6. The Office host application returns the **access token** to the add-in as part of the result object returned by the `getAccessToken` call.
7. The token is both an **access token** and an **identity token**. You can use it as an identity token to parse and examine claims about the user, such as the user's name and email address.
8. Optionally, the add-in can use the token as an **access token** to make authenticated HTTPS requests to APIs on the server-side. Because the access token contains identity claims, the server can store information associated with the user's identity; such as the user's preferences.

## Requirements and Best Practices

### Don't cache the access token

> Never cache or store the access token in your cient-side code. Always call [getAccessToken](/javascript/api/office-runtime/officeruntime.auth#getAccessToken_options_) when you need an access token. Office will cache the access token (or request a new one if it expired.) This will help to avoid accidentally leaking the token from your add-in.

Reviewer: ```suggestion
Never cache or store the access token in your cient-side code. Always call [getAccessToken](/javascript/api/office-runtime/officeruntime.auth#getAccessToken_options_) when you need an access token. Office caches the access token (or request a new one if it expired.) This helps to avoid accidentally leaking the token from your add-in.
```

## Develop an SSO add-in

This section describes the tasks involved in creating an Office Add-in that uses SSO. These tasks are described here independently of language or framework. For step-by-step instructions, see:

- [Create a Node.js Office Add-in that uses single sign-on](create-sso-office-add-ins-nodejs.md)
- [Create an ASP.NET Office Add-in that uses single sign-on](create-sso-office-add-ins-aspnet.md)

> [!NOTE]
> You can use the Yeoman generator to create an SSO-enabled, Node.js Office Add-in. The Yeoman generator simplifies the process of creating an SSO-enabled add-in, by automating the steps required to configure SSO within Azure and generating the code that's necessary for an add-in to use SSO. For more information, see the [Single sign-on (SSO) quick start](../quickstarts/sso-quickstart.md).

### Register your add-in with the Microsoft identity platform

> To work with SSO you need to register your add-in with the Microsoft identity platform. This will enable the Microsoft identity platform to provide authentication and authorization services for your add-in. Creating the app registration includes the following tasks.

Reviewer: ```suggestion
To work with SSO you need to register your add-in with the Microsoft identity platform. This enables the Microsoft identity platform to provide authentication and authorization services for your add-in. Creating the app registration includes the following tasks.
```

> [!IMPORTANT]
> As a best security practice, always call `getAccessToken` when you need an access token. Office will cache it for you. Don't cache or store the access token using your own code.

### Pass the access token to server-side code

If you need to access web APIs on your server, or additional services such as Microsoft Graph, you'll need to pass the access token to your server-side code. The access token provides access (for the authenticated user) to your web APIs. Also the server-side code can parse the token for identity information if it needs it. (See **Use the access token as an identity token** below.) There are many libraries available for different languages and platforms that can help simplify the code you write. For more information, see [Overview of the Microsoft Authentication Library (MSAL)](/azure/active-directory/develop/msal-overview).

If you need to access Microsoft Graph data, your server-side code should do the following:

- Validate the access token (see **Validate the access token** below).

> - Initiate the [OAuth 2.0 On-Behalf-Of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) with a call to the Microsoft identity platform that includes the access token, some metadata about the user, and the credentials of the add-in (its ID and secret). The Microsoft identity platform will return a new access token that can be used to access Microsoft Graph.

Reviewer: ```suggestion
- Initiate the [OAuth 2.0 On-Behalf-Of flow](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow) with a call to the Microsoft identity platform that includes the access token, some metadata about the user, and the credentials of the add-in (its ID and secret). The Microsoft identity platform returns a new access token that is used to access Microsoft Graph.
```
